third_party/blink/renderer/core/inspector/inspector_dom_agent.cc

```diff
@@ -767,33 +767,37 @@ int InspectorDOMAgent::PushNodePathToFrontend(Node* node_to_push,
   // InspectorDOMAgent might have been resetted already. See crbug.com/450491
   if (!document_)
     return 0;
-  if (!document_node_to_id_map_->Contains(document_))
-    return 0;
-
-  // Return id in case the node is known.
-  int result = node_map->at(node_to_push);
-  if (result)
-    return result;
 
-  Node* node = node_to_push;
-  HeapVector<Member<Node>> path;
-
-  while (true) {
-    Node* parent = InnerParentNode(node);
-    if (!parent)
-      return 0;
-    path.push_back(parent);
-    if (node_map->at(parent))
-      break;
-    node = parent;
-  }
-
-  for (int i = path.size() - 1; i >= 0; --i) {
-    int node_id = node_map->at(path.at(i).Get());
-    DCHECK(node_id);
-    PushChildNodesToFrontend(node_id);
-  }
-  return node_map->at(node_to_push);
+  // [replay] PushNodePathToFrontend is bad when `DevToolsSession` not active
+  //    TODO: This does not handle dangling nodes properly - https://linear.app/replay/issue/RUN-1005/
+  return BindDocumentNode(node_to_push);
+
+  // if (!document_node_to_id_map_->Contains(document_))
+  //   return 0;
+  // // Return id in case the node is known.
+  // int result = node_map->at(node_to_push);
+  // if (result)
+  //   return result;
+
+  // Node* node = node_to_push;
+  // HeapVector<Member<Node>> path;
+
+  // while (true) {
+  //   Node* parent = InnerParentNode(node);
+  //   if (!parent)
+  //     return 0;
+  //   path.push_back(parent);
+  //   if (node_map->at(parent))
+  //     break;
+  //   node = parent;
+  // }
+
+  // for (int i = path.size() - 1; i >= 0; --i) {
+  //   int node_id = node_map->at(path.at(i).Get());
+  //   DCHECK(node_id);
+  //   PushChildNodesToFrontend(node_id);
+  // }
+  // return node_map->at(node_to_push);
```

Comment:
This commented out code should be removed.

Reply:
This is not our code. That is why I kept it :thinking: 
Should we just delete their code?

Reply:
Hmm, looking more closely this change should be revised.  We shouldn't be making major changes to the browser's regular behavior like this, especially when those changes happen all the time and not when recording/replaying.  If we're not careful then changes can cause things in the browser to break in ways that will be extremely difficult to track down.

Can we use the new behavior only when recordreplay::HasDivergedFromRecording ?  Then existing inspector usage will not be affected and we'll only change what the browser does here when inspecting the DOM in replay pauses.

Reply:
I am not sure what requires `HasDivergedFromRecording`? But:

1. This code is only used during replay and
2. All these `InspectorXAgent`s (`DOM`, `CSS`, `Network`, `DOMDebug` etc.) are usually all part of a `DevToolsSession` and we can't really have a `DevToolsSession` active to begin with, so these paths would (I want to say with 99% certainty) only be used by us.

Also, in this case, I primarily remove behavior which is unlikely to be an issue. The new code (`BindDocumentNode`) is admittedly a hackfix. It is to make sure that all nodes have been "bound" (have an associated id) by the time we want to reference them (and afaik `PushNodePathToFrontend` generally has the same goal of making sure the frontend knows about any node that is being referenced in a CDP call).

The one major complication I could see as a result of this code is from "dangling nodes", and I am tracking this in https://linear.app/replay/issue/RUN-1005/dom-feature-test-and-support-dangling-dom-node-inspection.

Do you think this is Ok, or should we not ever even try to do this? What else do you suggest?

Reply:
Checking recordreplay::HasDivergedFromRecording() is essentially testing for whether we are replaying and at a pause where we are inspecting the application's state -- we diverge so we can start inspecting the state without worrying about side effects, and will not resume running through the recording later.

I don't understand what you mean by "this code is only used during replay" because the code we're discussing originates from upstream so it certainly can be used in other cases.

Removing or changing behavior needed by other clients using the CDP is what we want to avoid.  While it's unusual for someone to have the debugger open while creating a recording, it's something we want to support longer term.  More importantly, the CDP is used to drive automation, testing infrastructure etc. and we want that to work the same way whether the browser is being recorded.  We've already had problems with divergences in behavior, e.g. causing ebay's tests to hang for no apparent reason.

Reply:
Ok, I shall stop assuming that `chromium` usage behavior is somehow dictated by us, and assume that, ideally, all upstream code can still be used in the future. (I simply assumed, we can for now ignore the possibility of active `DevToolsSession` as future-work and kick the can down the road. In that case "this code is (very likely) only used during replay".)

I changed it back so that the agent still works as usual.
In order to allow execution without an active session during replay, some `if (enabled...)` checks are now effectively `if (enabled... || HasDivergedFromRecording())`. If code that requires the agent to be `enabled` (such as `PushNodePathToFrontend`) is executed, I handle it with a `if (!enabled && HasDivergedFromRecording())` hackfix.

Would that make sense @bhackett1024?

Reply:
Yes, this sounds good.